### PR TITLE
be helpful when throwing ConcurrentModificationException out of consumers

### PR DESCRIPTION
**Pull request by radai-rosenblatt**
Signed-off-by: Radai Rosenblatt <rrosenbl@rrosenbl-ld1.linkedin.biz>

instead of capturing the threadId of the thread currently accessing the consumer, capture
the thread itself. this means any exceptions thrown on concurrent use can be a lot more helpful
by pointing out the actual threads involved, instead of sending users to go fish

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 1]
```diff
@@ -589,7 +589,8 @@
 
     // currentThread holds the threadId of the current thread accessing KafkaConsumer
     // and is used to prevent multi-threaded access
-    private final AtomicLong currentThread = new AtomicLong(NO_CURRENT_THREAD);
+    private final AtomicLong currentThreadId = new AtomicLong(NO_CURRENT_THREAD);
```
Comment (smccauliff) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Is there a need to keep the thread id around if the Thread is already being referenced?

Comment (radai-rosenblatt) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: fixed

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -2419,18 +2417,22 @@ private void acquireAndEnsureOpen() {
...
     }
 
     /**
      * Release the light lock protecting the consumer from multi-threaded access.
      */
     private void release() {
-        if (refcount.decrementAndGet() == 0)
-            currentThread.set(NO_CURRENT_THREAD);
+        if (refcount.decrementAndGet() == 0) {
+            ownerThread.set(null);
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Should we use "NO_CURRENT_THREAD" to replace "null"? It seems to me "NO_CURRENT_THREAD" is more readable.

Comment (radai-rosenblatt) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: fine by me

**Comment by radai-rosenblatt** [comment]
@ijuma - look ok to you ?

**Comment by ijuma** [comment]
Thanks for the PR. Any concern regarding holding to the thread from a memory retention perspective?

**Comment by radai-rosenblatt** [comment]
holding to the Thread only becomes a problem if this is the last reference preventing the thread from being collected. that would mean a few things:

- thread has managed to die during a kafka call?
- (alternatively) we have a bug where finally wasnt called to release the lock
- this client is no longer usable as its owned by a thread that will never come back to relese it

i assume this doesnt happen, because if this was a realistic scenario the existing schema (that only captures thread is) would exhibit at least the 3rd issue (clients remaining locked forever).

so all in all @ijuma i think this is safe, as the consequences of holding on to the thread would be worse than a memory leak and would already be happening if this was realistic

**Comment by radai-rosenblatt** [comment]
@ijuma - any further modifications required?

**Comment by chia7712** [comment]
close as this is implemented

**Pull request closed without merging**